cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+                if any_overlap(&si, offset, size) {
+                    self.map[i] = None;
+                }
+            }
+        }
+    }
+
+    // Invalidates all bindings in the enviroment.
+    fn invalidate_all(&mut self) {
+        for i in 0..self.size {
+            self.map[i] = None;
+        }
+    }
+
+    // Updates the environment to track the effect of an `InstructionData::RegMove` instruction.
+    // This moves a value from one register to another, but doesn't create a new SSA name.  Hence
+    // we copy `src`'s entry to `dst` verbatim, without installing a new value in `dst`'s `Value`
+    // field.
+    fn move_reg(&mut self, src: RegUnit, dst: RegUnit) {
+        self.map[dst as usize] = self.map[src as usize];
+        self.map[src as usize] = None;
+    }
+
+    // Does `env` have the exact binding characterised by `(ref, offset, size)` ?
+    fn has_exact_binding(&self, reg: RegUnit, offset: i32, size: u32) -> bool {
+        if let Some(si) = &self.map[reg as usize] {
+            return si.offset == offset && si.size == size;
+        }
+        // No such binding.
+        false
+    }
+
+    // Does `env` have a binding characterised by `(offset, size)` but to a register, let's call
+    // it `reg2`, that isn't `reg`?  If so, return `reg2` and its associated SSA name.  Note that
+    // `reg2` will have the same bank as `reg`.  It is a checked error to call this function with
+    // a binding matching all three of `(reg, offset, size)`.
+    fn has_inexact_binding(
+        &self,
+        reginfo: &RegInfo,
+        reg: RegUnit,
+        offset: i32,
+        size: u32,
+    ) -> Option<(RegUnit, Value)> {
+        // Find the range of RegUnit numbers for the bank that contains `reg`, and use that as our
+        // search space.  This is so as to guarantee that any match is restricted to the same bank
+        // as `reg`.
+        let (first_unit, num_units) = find_bank_limits(reginfo, reg);
+        for reg2 in first_unit..first_unit + num_units {
+            if let Some(si) = &self.map[reg2 as usize] {
+                if si.offset == offset && si.size == size {
+                    if reg2 == reg {
+                        panic!("has_inexact_binding: binding *is* exact!");
+                    }
+                    return Some((reg2, si.value));
+                }
+            }
+        }
+        // No such binding.
+        None
+    }
+
+    // Create the binding `(reg, value, offset, size)` in `env`, and throw away any previous
+    // binding associated with `reg`.
+    fn bind(&mut self, reg: RegUnit, value: Value, offset: i32, size: u32) {
+        debug_assert!(self.map.len() == self.size);
+        self.map[reg as usize] = Some(SlotInfo {
+            value,
+            offset,
+            size,
+        });
+    }
+}
+
+impl<'a> Context<'a> {
+    /// Run the remover.
+    fn cleanup_func(&mut self) {
+        let mut total_regunits = 0;
+        for ru in self.reginfo.banks {
```

Comment:
nit: it's not clear what the U in "ru" stands for here (if that's for "units", then it's incorrect since we're iterating over register banks). Maybe the right short name is "rb"?